### PR TITLE
feat: add timestamp support for sqlx+postgres integration

### PR DESCRIPTION
**Pull request by cylewitruk**
## Description

Adds a new `Timestamp` model which can be seamlessly used in `sqlx` queries against `TIMESTAMPZ` columns.

## Changes

- New `Timestamp` type and related `sqlx` implementations.
- `Dummy` implementation which generates timestamps valid for PostgreSQL
- Integration test to verify it works

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/src/storage/sqlx.rs [hunk 1]
```diff
@@ -330,3 +343,68 @@ impl sqlx::postgres::PgHasArrayType for SigHash {
...
+impl<'r> sqlx::Decode<'r, sqlx::Postgres> for Timestamp {
+    fn decode(value: sqlx::postgres::PgValueRef<'r>) -> Result<Self, BoxDynError> {
+        // Decode the i64 representing microseconds since PostgreSQL epoch.
+        let pg_epoch_micros = <i64 as sqlx::Decode<sqlx::Postgres>>::decode(value)?;
+
+        // Convert microseconds since PostgreSQL epoch to seconds since Unix
+        // epoch.
+        // SAFETY: MICROS_PER_SECOND is a non-zero constant, so division by zero
+        // is not possible here.
+        let unix_seconds = (pg_epoch_micros / MICROS_PER_SECOND)
```
[the remark this answers is not in the document]
Comment (MCJOHN974) [reply signer/src/storage/sqlx.rs:391]: I think 1ms is 100% less then message from signer to signer or from whoever to signer so we ok with millis, but I don't see any damage from storing micros or nanos, we will make our db a bit bigger but 1 more byte per row sounds acceptable

[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/storage/sqlx.rs:391]: I don't see why we wouldn't support microseconds if we can. There is no cost to doing so, so why not.

[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/storage/sqlx.rs:391]: > but I don't see any damage from storing micros or nanos, we will make our db a bit bigger but 1 more byte per row sounds acceptable

I don't think it makes the database any larger. I might be mistaken but I think Postgres stores timestamps using 8 bytes, regardless of the level of precision that we give it.

File: signer/src/storage/model.rs [hunk 2]
```diff
@@ -1504,6 +1504,30 @@ pub struct BitcoinBlockHeight(u64);
 #[serde(transparent)]
 pub struct StacksBlockHeight(u64);
 
+/// A newtype over [`time::OffsetDateTime`] which implements encode/decode for sqlx
+/// and integrates seamlessly with the Postgres `TIMESTAMPTZ` type.
+#[derive(Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/storage/model.rs:1509]: No thanks, if we don't need to implement a trait, I see no good reason to do so.

File: signer/src/storage/sqlx.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/sqlx.rs:391]: Yeah, postgres stores them as 8 bytes and microseconds. Only reason it's seconds right now is that I originally had a `UnixTimestamp(i64)` type (UNIX timestamps are at second-resolution) instead of `Timestamp(time::OffsetDateTime)` (which supports down to nanos), and I just kept the existing math.

File: signer/src/storage/sqlx.rs [hunk 3]
```diff
@@ -330,3 +343,68 @@ impl sqlx::postgres::PgHasArrayType for SigHash {
...
+        std::mem::size_of::<i64>()
+    }
+}
+
+impl<'r> sqlx::Decode<'r, sqlx::Postgres> for Timestamp {
+    fn decode(value: sqlx::postgres::PgValueRef<'r>) -> Result<Self, BoxDynError> {
+        // Decode the i64 representing microseconds since PostgreSQL epoch.
+        let pg_epoch_micros = <i64 as sqlx::Decode<sqlx::Postgres>>::decode(value)?;
+
+        // Convert microseconds since PostgreSQL epoch to seconds since Unix
```
Comment (djordon) [inline signer/src/storage/sqlx.rs:387]: Do we need to worry about the format of the value returned from postgres? The sqlx implementation suggests that we should, but I'm not sure. https://github.com/launchbadge/sqlx/blob/bab1b022bd56a64f9a08b46b36b97c5cff19d77e/sqlx-postgres/src/types/time/datetime.rs#L54-L89

Comment (cylewitruk) [reply signer/src/storage/sqlx.rs:387]: Not according to the integration test, anyway. It looks like their implementation handles conversion from textual format, etc. We could always add that, but `TIMESTAMPZ` itself should be stored as an `i64` iiuc.

File: signer/src/storage/sqlx.rs [hunk 4]
```diff
@@ -23,6 +24,18 @@ use crate::storage::model::StacksBlockHeight;
...
+
+/// Seconds between 1970-01-01 00:00:00 UTC (Unix epoch)
+/// and 2000-01-01 00:00:00 UTC (PostgreSQL epoch for TIMESTAMPTZ).
+const PG_EPOCH_SECONDS_FROM_UNIX_EPOCH: i64 = 946_684_800;
+
+/// Number of microseconds in a second (as an `i64`).
+const MICROS_PER_SECOND: i64 = 1_000_000;
+
+/// OID for PostgreSQL's TIMESTAMPTZ type.
+const TIMESTAMPTZ_OID: Oid = Oid(1184);
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/sqlx.rs:37]: Kindof a shame that sqlx decided to keep a lot of these things private...

File: signer/src/storage/sqlx.rs [hunk 5]
```diff
@@ -23,6 +24,18 @@ use crate::storage::model::StacksBlockHeight;
 use crate::storage::model::StacksPrincipal;
 use crate::storage::model::StacksTxId;
 
+use super::model::Timestamp;
+
+/// Seconds between 1970-01-01 00:00:00 UTC (Unix epoch)
+/// and 2000-01-01 00:00:00 UTC (PostgreSQL epoch for TIMESTAMPTZ).
+const PG_EPOCH_SECONDS_FROM_UNIX_EPOCH: i64 = 946_684_800;
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/sqlx.rs:31]: Not quite sure what you're asking? The value is the unix timestamp of the PG epoch, which is in seconds. To get the postgres microsecond representation you'd multiply it by `MICROS_PER_SECOND` (1 million).

[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/storage/sqlx.rs:31]: I was asking you to add a link in the comment of the variable that shows that the PG epoch is 2000-01-01 00:00:00 UTC.

File: signer/src/storage/sqlx.rs (hunk 3 above)
Comment (djordon) [reply signer/src/storage/sqlx.rs:387]: It's weird that the sqlx folks support whatever it is that they support there; it's as if they support going from a varchar to a timestamp, but that cannot be what they are actually supporting. Hmmm, seems like it's something to keep in mind.

File: signer/src/storage/sqlx.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/storage/sqlx.rs:37]: Yeah, 😞 it's a shame. This is not the first time that I've thought that about sqlx.

File: signer/src/storage/sqlx.rs (hunk 3 above)
Comment (djordon) [reply signer/src/storage/sqlx.rs:387]: Resolving since there isn't really anything actionable.

File: signer/src/storage/sqlx.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/sqlx.rs]: Hard to find :) But added a link to wikipedia where it's mentioned in [f1b6a80](https://github.com/stacks-sbtc/sbtc/pull/1682/commits/f1b6a80e9861dae23953a8bb602b68d18e8c71c1)

File: signer/src/storage/sqlx.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/sqlx.rs]: Added the above link in [65c7013](https://github.com/stacks-sbtc/sbtc/pull/1682/commits/65c7013a3cd24db49f9722ccd4d3e9e2cfcb6390)

File: signer/src/storage/model.rs [hunk 6]
```diff
@@ -1504,6 +1504,30 @@ pub struct BitcoinBlockHeight(u64);
...
+pub struct Timestamp(time::OffsetDateTime);
+
+impl std::fmt::Debug for Timestamp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/model.rs]: It was rather to avoid getting `Timestamp(OffsetDateTIme(xxxxxx....))`, but I'm fine just using the default derive and changing it later if we really want to.
[fc7da19](https://github.com/stacks-sbtc/sbtc/pull/1682/commits/fc7da197e36d77d0067b232c76c0dbb842f4245b)

File: signer/src/storage/sqlx.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/sqlx.rs]: This specific constant was removed in place of another in [977c116](https://github.com/stacks-sbtc/sbtc/pull/1682/commits/977c116239553f8cfeac86bbf5af48f51ce62b7c), but I retained the wikipedia link.

File: signer/src/storage/sqlx.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/sqlx.rs]: Changed it to microsecond precision in [977c116](https://github.com/stacks-sbtc/sbtc/pull/1682/commits/977c116239553f8cfeac86bbf5af48f51ce62b7c)

**Review by MCJOHN974 — APPROVED** [review]

File: signer/src/storage/sqlx.rs (hunk 3 above)
Comment (djordon) [reply signer/src/storage/sqlx.rs]: Yeah, but I find that so strange. Wouldn't the OID read differently for TEXT data, or something like that? Well I don't understand how it all works so maybe it makes some sense.

File: signer/src/storage/sqlx.rs [hunk 7]
```diff
@@ -330,3 +342,64 @@ impl sqlx::postgres::PgHasArrayType for SigHash {
...
+        // Add this duration to the PostgreSQL epoch datetime.
+        // checked_add handles potential overflow/underflow if the resulting datetime
+        // is outside the representable range of OffsetDateTime.
+        let datetime = POSTGRES_EPOCH_DATETIME
+            .checked_add(duration_from_pg_epoch)
+            .ok_or_else(|| {
+                sqlx::Error::Decode(
+                    "failed to construct OffsetDateTime from decoded TIMESTAMPTZ value due to overflow/underflow".into(),
+                )
+            })?;
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/sqlx.rs]: [f23e209](https://github.com/stacks-sbtc/sbtc/pull/1682/commits/f23e209b30d4fac1fd8221b1b291504d3e853782)

**Review by MCJOHN974 — APPROVED** [review]

File: signer/src/testing/dummy.rs [hunk 8]
```diff
@@ -771,6 +771,22 @@ impl fake::Dummy<fake::Faker> for BitcoinPreSignAck {
...
+
+impl fake::Dummy<fake::Faker> for model::Timestamp {
+    fn dummy_with_rng<R: rand::RngCore + ?Sized>(_: &fake::Faker, rng: &mut R) -> Self {
+        // The PostgreSQL epoch is 2000-01-01 00:00:00 UTC
+        const PG_UNIX_EPOCH: i64 = 946_684_800;
+        // Let's try to be somewhat realistic: 2050-01-01 00:00:00 UTC
+        const TIMESTAMP_MAX: i64 = 2_524_608_000;
+        // Generate a random timestamp between the PostgreSQL epoch and TIMESTAMP_MAX (2050-01-01 00:00:00 UTC)
+        // to avoid PG overflow.
+        let unix_timestamp: i64 = rng.gen_range(PG_UNIX_EPOCH..TIMESTAMP_MAX);
```
Comment (matteojug) [inline signer/src/testing/dummy.rs:783]: qq: should we generate also dates from before `PG_UNIX_EPOCH` to test negative stuff as well? I don't see any possible usage of those, nevertheless why not
